# -*- coding: utf-8 -*-
"""Assignment 3.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/14g7FGSYgiXh7m3_I1ocRyxtFaJPgY7OP

Q1: Assume that we have one DNA sequence that has a string as

'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
which is comprised of four nucleotides (e.g. A, T, C, G). We would like to define a variable called 'DNA', and assign the sequence to this variable. Then we need to print the variable by print("DNA: ", DNA)
Please write your python codes for this activity.

How long is this sequence?
What is the first letter of the sequence?
What is the last letter of the sequence?
What are the first 10 characters of the sequence?
What are the last 10 characters of the sequence?
How many times does the character 'C' appear in the sequence?
What is the index at which the first occurrence of the pattern CCAAT begins? If not found, print out any warning message (i.e., The motif is not found)


Note: This DNA sequence will be applied to all the following questions.
"""

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
print(len(DNA))
print(DNA[0])
print(DNA[-1])
print(DNA[0:10])
print(DNA[10:0:-1])
print(DNA.count('C'))
print(DNA.count('CCAAT'))

"""Q2: To check if two Python values, or variables, are equal you can use ==. To check for inequality, you need !=.  As a refresher, have a look at the following examples that all result in True. Feel free to try them out in the Notebook.

2 == (1 + 1)
"intermediate" != "python"
True != False
"Python" != "python"
In the notebook, you need to wrap a print() (Links to an external site.) function around them to see the output.

Based on the sequence defined in Q1, write codes to see if the first 10 characters of the sequence and the last 10 characters of the sequence are equal.

Write Python code to check if the number of the character 'C' is equal to the number of the character 'A' in the DNA sequence.
"""

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
if DNA[0:10] == DNA[10:0:-1]:
  print("True")
else:
  print("False")
if DNA.count('C') == DNA.count('A'):
  print("True")
else:
  print("False")

"""Q3: Greater and less than

In python,  the comparison operators also include the less than and greater than signs, < and > in Python. You can combine them with an equals sign: <= and =>. Pay attention: <= is valid syntax, but =< is not. All Python expressions in the following code chunk evaluate to True:

3 < 4
3 <= 4
In the notebook, you need to wrap a print() (Links to an external site.) function around them to see the output.

Write Python code to check if the number of the character 'C'  is greater than or equal to the number of the character 'A' in the DNA sequence.

Write Python code to check if the index at which the first occurrence of the letter 'C' is greater than the index at which the first occurrence of the letter 'A'.
"""

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
if DNA.count('C') >= DNA.count('A'):
  print("C is greater or equal to A")
else:
  print("C is not great or equal to A")
if DNA.index('C') > DNA.index('A'):
  print("index of C is greater to A")
else:
  print("index of C is not greater to A")

"""Q4: A boolean is True or False. With boolean operators such as and, or and not, you can combine these booleans to perform more advanced queries on your data.

In the notebook, you need to wrap a print() (Links to an external site.) function around them to see the output.

Write Python code to check if the number of the character 'C'  is greater than the number of the character 'A'  and smaller than the number of the character 'G' in the DNA sequence.
Write Python code to check if the number of the character 'C'  is greater than the number of the character 'A'  or smaller than the number of the character 'G' in the DNA sequence.
Suppose  we have variables x = 8  and y = 9, what will the result be if you execute the command not(not(x < 3) and not(y > 14 or y > 10))  in the notebook?   Note: not has higher precedence than and  and or.
"""

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
if DNA.count('C') > DNA.count('A') and DNA.count('G'):
  print("C is greater than A and smaller than G")
else:
  print("C is not greater then A and smaller than G")

if DNA.count('C') > DNA.count ('A') or DNA.count('G'):
  print("C is greater than A or smaller than G")
else:
  print("C is not greater then A or smaller than G")

x = 8
y = 9
not(not(x < 3) and not(y > 14 or y > 10))

"""Q5: To experiment with if and else a bit, have a look at this code sample:

area = 10.0
if(area < 9) :
    print("small")
elif(area < 12) :
    print("medium")
else :
    print("large")
What will the output be if you run this piece of code in the notebook?
Use if, elif and else, check the length of DNA sequence in Q1 with the following conditions: if the length is shorter than 100, print 'the DNA sequence is short.', otherwise if the length is shorter than 500, print "the DNA sequence is medium", otherwise, print "the length is long".
"""

area = 10.0
if(area < 9) :
    print("small")
elif(area < 12) :
    print("medium")
else :
    print("large")

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
if len(DNA) < 100:
  print("The DNA sequence is short")
elif len(DNA) < 500:
  print("The DNA sequence is medium")
else:
  print("The length is long")

"""Q6: The while loop is like a repeated if statement. The code is executed over and over again, as long as the condition is True. Have another look at its recipe.

while condition :
    expression
have a look at this code sample:

x = 1
while x < 4 :
    print(x)
    x = x + 1
Can you tell how many printouts the above while loop will do?
Use while loop, keep printing the letters from left to right in the DNA sequence until you reach the first letter 'C'. Note: recall how the keyword break or continue works.
Use while loop, keep printing the letters from left to right in the DNA sequence if the letter is not 'C'. Note: recall how the keyword break or continue works.
Use while loop, keep printing the letters from left to right in the DNA sequence if the letters are not 'C' and 'G'. Note: recall how the keyword break or continue works.
Use while loop, keep printing the letters from left to right in the DNA sequence if the letters are either 'A' or 'G'. Note: recall how the keyword break or continue works.

"""

x = 1
while x < 4 :
    print(x)
    x = x + 1

DNA = 'GTTGATGTAGCTTATATAAAGCAAGGCACTGAAAATGCCTAGATGAGTCATAGACTCCATAAACAACAGGTTTGGTCCCGGCCTTTTTATTAGCTGTCTGCAGGATTATACATGCAAAAATCCCTACACCGGTGAGAATGCCCTATGTACCACACTCAGGTTTAAAGGAGCAGACATCAAGCACACTGCTAAGTAGCTCACGACGTCTTGCTT'
n = 0
nuc = ''
while DNA[n] != 'C':
  nuc += (DNA[n])
  n = n + 1
print (nuc)

n = 0
nuc = ''
while n < len(DNA):
  if DNA[n] == 'C':
    n = n + 1
    continue
  else:
    nuc += (DNA[n])
    n = n + 1
print(nuc)

n = 0
nuc = ''
while n < len (DNA):
  if DNA[n] == 'C' or DNA[n] == 'G':
    n = n + 1
    continue
  else:
    nuc += (DNA[n])
    n = n + 1
print(nuc)

n = 0
nuc = ''
while n < len(DNA):
  if DNA[n] == 'A' or DNA[n] == 'G':
    n = n + 1
    continue
  else:
    nuc += (DNA[n])
    n = n + 1
print(nuc)

"""Q7: have a look at this code sample:

course_ids= ['CSCI1020', 'CSCI1030', 'CSCI1070'] for id in course_ids: print(id)
As usual, you simply have to indent the code with 4 spaces to tell Python which code should be executed in the for loop.

The course_ids variable, containing the ids of different courses in the Computer Science Department, is already defined.

Write a for loop that goes through each element of course_ids. On each iteration, "the course id is x" should be printed out, where x is the id value in this iteration.
Use for loop, keep printing the letters from left to right in the DNA sequence until you reach the letter 'C'.  Note: recall how the keyword break or continue works
Use for loop, keep printing the letters from left to right in the DNA sequence if the letters are either 'A' or 'G'. Note: recall how the keyword break or continue works.
Note well that we do an index-based loop with for ind in range(0,length). As a sanity check, if the dna has length 5, given for ind in range(0,5), we will test five elements: dna[0], dna[1],  dna[2],  dna[3],  and dna[4].
"""

course_ids = ['CSCI1020', 'CSCI1030', 'CSCI1070']
for id in course_ids:
  print(course_ids)

for course in course_ids:
  print("the course id is", course)

print("question2 ")
for motif in DNA:
  print(motif, end = '')
  if motif == 'C':
    break
print(" question 3")
for motif in DNA:
  print(motif, end = '')
  if motif == 'A' or "G":
    break

"""Q8: In Python, you need the items() (Links to an external site.) method to loop over a dictionary:

CS_courses= { "CSCI1020":"Intro to bioinformatics",
          "CSCI1030":'Intro to Game Design',
          "CSCI1070":'Intro  to Big Data' }

for key, value in CS_courses.items() :
    print(key + " -- " + str(value))
Remember the CS_courses dictionary that contained the names of some course ids as key and their course names as corresponding value? Go ahead and write a loop to iterate over it

Write a for loop that goes through each key:value pair of CS_courses. On each iteration, "the course name of x is y" should be printed out, where x is the key and y is the value of the pair.
"""

CS_courses= { "CSCI1020":"Intro to bioinformatics",
          "CSCI1030":'Intro to Game Design',
          "CSCI1070":'Intro  to Big Data' }

for key, value in CS_courses.items() :
    print(key + " -- " + str(value))

"""Q9: According to the DNA sequence defined in Q1, write your Python codes to answer the following questions:
1. How often is a letter followed immediately by the same letter in the DNA sequence? Note well that we do an index-based loop but with range(len(dna)-1). As a sanity check, if the dna has length 5, we only need to test four pairs of neighbors, and thus compare dna[0] to dna[1], dna[1] to dna[2], dna[2] to dna[3], and dna[3] to dna[4].
2. If two consecutive nucleotides match each other, how often is the next nucleotide the same nucleotide?

3. How many times does a motif of form CC?AT occur within the sequence? (where ? could be anything)

4. When the motif CC?AT does occur, what percentage of the time is the middle nucleotide an A? (A so-called cat box CCAAT)
"""

index1 = 0
index2 = 1
num_of_repeats = 0
while index2 < len(DNA):
  if DNA[index1] == DNA[index2]:
    num_of_repeats += 1
  index1 += 1
  index2 += 1
print(f'a letter is followed immediately by the same letter {num_of_repeats} times')

index1 = 0
index2 = 1
index3 = 2
num_of_repeats = 0
while index3 < len(DNA):
  if DNA[index1] == DNA[index2] == DNA[index3]:
    num_of_repeats += 1
  index1 += 1
  index2 += 1
  index3 += 1
print(f'there are {num_of_repeats} instances where two consecutive matching nucleotides are followed by another matching nucleotide')

num_of_instances = 0
index = 0
while index < len(DNA) - 5:
  if DNA[index:index + 5] == 'CCAAT' or DNA[index:index + 5] == 'CCCAT' or DNA[index:index + 5] == 'CCGAT' or DNA[index:index + 5] == 'CCTAT':
    num_of_instances += 1
  index += 1
print(f"the motif of form CC?AT occurs {num_of_instances} times")

num_of_instances = 0
index = 0
CCAAT = 0
CCCAT = 0
CCGAT = 0
CCTAT = 0
while index < len(DNA) - 5:
  if DNA[index:index + 5] == 'CCAAT':
    CCAATT += 1
  if DNA[index:index + 5] == 'CCCAT':
    CCCAT += 1
  if DNA[index:index + 5] ==  'CCGAT':
    CCGAT += 1
  if DNA[index:index + 5] == 'CCTAT':
    CCTAT += 1
  index += 1
percent_CCAAT = CCAAT/(CCCAT + CCGAT + CCTAT)
print(f'percent of CC?AT that is CCAAT: {percent_CCAAT} percent')

"""Q10: In this part, we will extend question 11 of our lab2 to explore the existence of inverted repeats (Links to an external site.) in any DNA sequence, let's start again with the background:

Background: The DNA sequence contains inverted repeats when some particular motif is later followed in a DNA strand by its reverse complement. In three dimensions, such a single strand might recombine in a loop, with the motif binding with its inverted repeat, forming what is known as a transposable element (Links to an external site.) which often is a source of mutations.

As an example, consider the single strand ATCCGAATACGGTTCGGGTA. Notice the motif CGAA that is later followed by the reverse complement TTCG. While this is only one strand of what is double-stranded DNA, during the replication process when it is single-stranded, there is a chance that it binds with itself at the motif and its inverted repeat, forming a loop as follows:

...ATC GGTA...
     C G
     G C
     A T
     A T
    T   G
     ACG
If that loop were to be disconnected for some reason, the result mutation would result in the remaining ATCGGTA.
"""

def reverse_complement(dna):
  dna2dna = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}
  other = ''
  for base in dna:
    other = other + dna2dna[base]
  return other[ : : -1]

def mutate (DNA,motif):
  if motif not in DNA:
    return DNA
  else:
    if reverse_complement(motif) not in DNA:
      return DNA
    else:
      motif_index = DNA.find(motif)
      reverse_index = DNA.find(reverse_complement(motif)) + len(reverse_complement(motif))

      next_instance = 1
      while reverse_index < motif_index:
        reverse_index = DNA.find(reverse_complement(motif), DNA.find(reverse_complement(motif)) + next_instance) + len(reverse_complement(motif))
        next_instance += 1

      DNA = DNA[:motif_index] + DNA[reverse_index:]
      return DNA